File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minPq = new PriorityQueue<>(); // 오름차순
        PriorityQueue<Integer> maxPq = new PriorityQueue<>(Collections.reverseOrder()); // 내림 차순
        
        for(String op : operations){
            String[] array = op.split(" ");
            
            if(array[0].equals("I")){
                int num = Integer.parseInt(array[1]); // String -> int
                minPq.offer(num);
                maxPq.offer(num);
            } else{
                if(maxPq.isEmpty() && minPq.isEmpty()) {continue;}
                else {
                    if(array[1].equals("1")){ // 최댓값 삭제
                        int max = maxPq.poll();
                        minPq.remove(max);
                    } else if(array[1].equals("-1")){ // 최솟값 삭제
                        int min = minPq.poll();
                        maxPq.remove(min);
                    }
                }
            } 
        }
        
        int[] answer = new int[2];
        if(minPq.isEmpty() && maxPq.isEmpty()){
            answer[0] = 0;
            answer[1] = 0;
        } else {
            answer[0] = maxPq.peek();
            answer[1] = minPq.peek();
        }
        
        
        return answer;
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq as hq

def solution(operations):
    answer = []
    heap = []
    
    for operation in operations:
        op, data = operation.split()

        if op == "I":
            hq.heappush(heap, int(data))
        elif op == "D" and data == "-1" and heap: # 최솟값 추출
            hq.heappop(heap)
        elif op == "D" and data == "1" and heap: # 최대값 추출
            heap.remove(max(heap))
            hq.heapify(heap)
        
        # print(heap)
        
    if heap:
        return [max(heap), min(heap)]
    else:
        return [0, 0]
    
    return answer